### PR TITLE
add support for opus files

### PR DESCRIPTION
**Pull request by zaethan**
This adds support for opus files.

It _seems_ to work fine on my limited set of test files. One thing I noticed is that sometimes a tag is written in uppercase "TITLE" (when set with kid3) and sometimes in lowercase "title" (when converted from mp3 with ffmpeg). To be case insensitive I used `to_uppercase()`.

Another thing I noticed is that the duration is not displayed. However it seems like it is not supposed to be part of the header but should rather be determined from the last ogg page of the stream.

### CODE REVIEW

**Comment by lnicola** [comment]
I think you can use https://doc.rust-lang.org/std/primitive.str.html#method.eq_ignore_ascii_case to avoid a couple of allocations.

**Comment by zaethan** [comment]
In that case, using the `match` would not be possible anymore, right?
It would have to be replaced with multiple `if "TAG".eq_ignore_ascii_case(key.as_str()) {}` statements.
I think the code is less readable this way, but if it is faster it seems to be a good idea nonetheless.

File: src/index/metadata.rs [hunk 1]
```diff
@@ -175,6 +177,42 @@ fn read_vorbis(path: &Path) -> Result<SongTags> {
...
+		album: None,
+		title: None,
+		duration: None,
+		disc_number: None,
+		track_number: None,
+		year: None,
+	};
+
+	for (key, value) in headers.comments.user_comments {
+		if "TITLE".eq_ignore_ascii_case(key.as_str()) {
```
Comment (lnicola) [inline src/index/metadata.rs:196]:
```suggestion
		if "TITLE".eq_ignore_ascii_case(&key) {
```

File: src/index/metadata.rs [hunk 2]
```diff
@@ -175,6 +177,42 @@ fn read_vorbis(path: &Path) -> Result<SongTags> {
...
+			tags.title = Some(value);
+		} else if "ALBUM".eq_ignore_ascii_case(key.as_str()) {
+			tags.album = Some(value);
+		} else if "ARTIST".eq_ignore_ascii_case(key.as_str()) {
+			tags.artist = Some(value);
+		} else if "ALBUMARTIST".eq_ignore_ascii_case(key.as_str()) {
+			tags.album_artist = Some(value);
+		} else if "TRACKNUMBER".eq_ignore_ascii_case(key.as_str()) {
+			tags.track_number = value.parse::<u32>().ok();
```
Comment (lnicola) [inline src/index/metadata.rs:205]:
```suggestion
			tags.track_number = value.parse().ok();
```

**Comment by lnicola** [comment]
There's a small thing you can do to make it easier on the eyes: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=717ab9c3845ae87f29342a5aed1f8537 (hmm, it might be worth putting that into a crate).

But I don't thing the if-else chain is so bad. It's pretty clear what it does, and I think it's worth avoiding an allocation for each comment.

**Comment by zaethan** [comment]
That macro seems to be a really good way to deal with this if-else chain. This way, it could easily be applied to the `read_vorbis` function as well, as the [vorbis spec](https://www.xiph.org/vorbis/doc/v-comment.html) states that it also uses case insensitive comments.
However, I don't really know where the best place to define macros is. Inside the file that uses it? In an extra module?

**Comment by lnicola** [comment]
Either way if fine, I think you can import it just like a function if it's in the same crate. Maybe you can make a small module for it, or put it in lib.rs.

File: Cargo.toml [hunk 3]
```diff
@@ -45,6 +45,7 @@ thiserror = "1.0.19"
 time = "0.1"
 toml = "0.5"
 url = "2.1"
+opus_headers = "0.1.2"
```
Comment (agersant) [inline Cargo.toml:48]: Can you move this up a bit so the libraries are still sorted alphabetically?

**Comment by agersant** [comment]
> Another thing I noticed is that the duration is not displayed. However it seems like it is not supposed to be part of the header but should rather be determined from the last ogg page of the stream.
There was a similar can of worms for other formats (hence the existence of `mp3-duration`). I think this is fine to defer, supporting metadata is already a good improvement!

The macro is very neat, this does make things more readable. I would suggest leaving it in the file that uses it, or alternatively in `utils.rs` which is a small dumping ground for code with no great home.

Can you also make use of your `sample.opus` test file in `test_read_metadata()` so it becomes part of the test suite?

**Comment by zaethan** [comment]
I put the macro at the top of the metadata.rs. Initially I wanted to put it into utils.rs, but I if I remember correctly, this implies a certain order of the `mod ...` statements in main.rs such that `mod utils` must be called before any mod that uses the macro.
As it has the same requirements for case insensitivity, I also applied the macro to the read_vorbis function.

I also included the sample.opus in the tests. To ensure that case insensitivity works, I changed the comments in the opus file to use varying casing, like "AlBuMaRtIsT" or "ARTIsT".

**Comment by lnicola** [comment]
> if I remember correctly, this implies a certain order of the mod ... statements in main.rs such that mod utils must be called before any mod that uses the macro

I don't think so, or at least I've never seen such a case in Rust.

**Comment by zaethan** [comment]
I just tried it out and putting the macro in the utils module causes a build error (cannot find macro `match_ignore_case` in this scope) and a warning (unused macro definition). Using `#[macro_use] mod util;` does not fix it. However, if I move `#[macro_use] mod util;` above `mod index` in the main.rs, it builds fine again.

Looking it up, I found a [reddit thread](https://www.reddit.com/r/rust/comments/2lc84e/importing_macros_from_another_module/), where someone linked [this](https://danielkeep.github.io/quick-intro-to-macros.html#some-more-gotchas) post about macro definitions. However, given that the reddit thread is 5 years old, it might be outdated...

**Comment by lnicola** [comment]
Does it work if you import it, e.g. `use crate::util::match_ignore_case`?

**Comment by zaethan** [comment]
Not directly. Adding this `use` leads to an `unresolved import crate::utils::match_ignore_case`, `no match_ignore_case in utils` error.
Trying around some more, I found that annotating the macro with `#[macro_export]` and then adding use `crate::match_ignore_case` (note the missing utils) works.
It seems like `#[macro_export]` "moves" the macro to the root of the crate...

**Comment by lnicola** [comment]
TIL: it seems that the "correct" incantation is:

```rust
#[macro_export]
macro_rules! match_ignore_case {
    (match $v:ident {
        $( $lit:literal => $res:expr, )*
        _ => $catch_all:expr $(,)?
    }) => {{
        $( if $lit.eq_ignore_ascii_case(&$v) { $res } else )*
        { $catch_all }
    }};
}

pub use crate::match_ignore_case;
```

**Comment by zaethan** [comment]
This is interesting. I think the way this works is that `#[macro_export]` makes the macro visible for the crate root and it can be accessed through `crate::macro`. Adding `pub use ...` then makes the macro accessible through the module foo, which is what we initially wanted to achieve.
I think we should do that, since it makes clear where the macro actually comes from.

**Comment by agersant** [comment]
Ship it! 🥳

**Pull request merged**